File: Net/02/TasksManager.Application/TaskTypesStatsEventHandler.cs
using Commons.Services;
using Commons.Tasks;
using Cqrs;
using log4net;
using System;
using Tasks.Repositories;
using TasksManager.Repositories;

namespace TasksManager
{
    public class TaskTypesStatsEventHandler : IMessageHandler
    {
        private static ILog _logger = LogManager.GetLogger(typeof(TaskTypesStatsEventHandler));
        private ITasksService _tasksService;
        private IRepository<TaskTypeStatDao, String> _repository;
        private IBus _bus;

        public void Register(IBus bus)
        {
            _bus = bus;
            _bus.RegisterHandler(m => Handle((TaskCompleted)m), typeof(TaskCompleted));
            _bus.RegisterHandler(m => Handle((TaskCreated)m), typeof(TaskCreated));
        }

        public TaskTypesStatsEventHandler(ITasksService tasksService, IRepository<TaskTypeStatDao, String> repository)
        {
            _repository = repository;
            _tasksService = tasksService;
        }

        public void Handle(TaskCompleted message)
        {
            _logger.Info("TaskCompleted");
            TaskDao taskDao = _tasksService.GetById(message.Id);
            String code = taskDao.TypeCode == null ? "NONE" : taskDao.TypeCode;

            TaskTypeStatDao currentStat = _repository.GetById(code);
            if (currentStat == null)
            {
                throw new IndexOutOfRangeException("Missing stats for code " + code);
            }

            currentStat.Completed = currentStat.Completed + 1;
            currentStat.Running = currentStat.Running - 1;

            _repository.Save(currentStat);
        }

        public void Handle(TaskCreated message)
        {
            _logger.Info("TaskCreated");
            TaskDao taskDao = _tasksService.GetById(message.Id);
            String code = taskDao.TypeCode == null ? "NONE" : taskDao.TypeCode;

            TaskTypeStatDao currentStat = _repository.GetById(code);
            if (currentStat == null)
            {
                currentStat = new TaskTypeStatDao();
                currentStat.TypeCode = code;
            }

            currentStat.Running = currentStat.Running + 1;

            _repository.Save(currentStat);
        }
    }
}

File: Java/02/tasks-manager-application/src/main/java/org/tasksmanager/TaskTypesStatsEventHandler.java
package org.tasksmanager;

import java.util.logging.Level;
import java.util.logging.Logger;
import javax.inject.Inject;
import javax.inject.Named;
import org.commons.Services.TaskDao;
import org.commons.Services.TasksService;
import org.commons.Tasks.TaskCompleted;
import org.commons.Tasks.TaskCreated;
import org.cqrs.Bus;
import org.cqrs.MessageHandler;
import org.cqrs.Repository;
import org.tasksmanager.Repositories.TaskTypeStatDao;

@Named("taskTypesStatsEventHandler")
public class TaskTypesStatsEventHandler implements MessageHandler {

    private static final Logger logger = Logger.getLogger(TaskTypesStatsEventHandler.class.getSimpleName());
    private final TasksService _tasksService;
    private final Repository<TaskTypeStatDao, String> _repository;
    private Bus _bus;

    @Override
    public void Register(Bus bus) {
        _bus = bus;
        _bus.RegisterHandler(m -> Handle((TaskCompleted) m), TaskCompleted.class);
        _bus.RegisterHandler(m -> Handle((TaskCreated) m), TaskCreated.class);
    }

    @Inject
    public TaskTypesStatsEventHandler(TasksService tasksService, Repository<TaskTypeStatDao, String> taskTypeStatsRepository) {
        _repository = taskTypeStatsRepository;
        _tasksService = tasksService;
    }

    public void Handle(TaskCompleted message) {
        logger.log(Level.INFO, "TaskCompleted");
        TaskDao taskDao = _tasksService.GetById(message.getId());
        String code = taskDao.getTypeCode() == null ? "NONE" : taskDao.getTypeCode();

        TaskTypeStatDao currentStat = _repository.GetById(code);
        if (currentStat == null) {
            throw new RuntimeException("Missing stats for code " + code);
        }

        currentStat.setCompleted(currentStat.getCompleted() + 1);
        currentStat.setRunning(currentStat.getRunning() - 1);

        _repository.Save(currentStat);
    }

    public void Handle(TaskCreated message) {
        logger.log(Level.INFO, "TaskCreated");
        TaskDao taskDao = _tasksService.GetById(message.getId());
        String code = taskDao.getTypeCode() == null ? "NONE" : taskDao.getTypeCode();

        TaskTypeStatDao currentStat = _repository.GetById(code);
        if (currentStat == null) {
            currentStat = new TaskTypeStatDao();
            currentStat.setTypeCode(code);
        }

        currentStat.setRunning(currentStat.getRunning() + 1);

        _repository.Save(currentStat);
    }
}
